src/rate-limit.coffee

```diff
@@ -53,27 +55,39 @@ module.exports = (robot) ->
       else
         roomOrUser = context.response.message.user.name
 
-      # Construct a key to rate limit on. If the response was from a room
-      # then append the room name to the key. Otherwise, append the user name.
-      listenerAndRoom = listenerID + "_" + roomOrUser
+      # Get the key used for storing data in lastExecutedTime and
+      # lastNotifiedTime
+      if context.listener.options.rateLimits &&
+         context.listener.options.rateLimits.keyFunction
+        # Get key by calling a user-defined function
+        robot.logger.debug "Getting key from keyFunction"
+        key = context.listener.options.rateLimits.keyFunction(context)
```

Comment:
I wonder if this should also include the `roomOrUser`. Without it, you are rate limiting globally instead of per conversation, which results in funky behavior (I direct message the bot and don't get a reply because someone else direct messaged a half second before me). Along those lines, I think the right answer here is to just take the existing key (`listenerAndRoom`) and append the custom key to the end.